File: back/src/main/java/com/tdevilleduc/urthehero/back/service/impl/ProgressionService.java
package com.tdevilleduc.urthehero.back.service.impl;

import com.tdevilleduc.urthehero.back.dao.ProgressionDao;
import com.tdevilleduc.urthehero.back.exceptions.PageNotFoundException;
import com.tdevilleduc.urthehero.back.exceptions.PersonNotFoundException;
import com.tdevilleduc.urthehero.back.exceptions.ProgressionNotFoundException;
import com.tdevilleduc.urthehero.back.exceptions.StoryNotFoundException;
import com.tdevilleduc.urthehero.back.model.Page;
import com.tdevilleduc.urthehero.back.model.Progression;
import com.tdevilleduc.urthehero.back.service.IPageService;
import com.tdevilleduc.urthehero.back.service.IPersonService;
import com.tdevilleduc.urthehero.back.service.IProgressionService;
import com.tdevilleduc.urthehero.back.service.IStoryService;
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.helpers.MessageFormatter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static com.tdevilleduc.urthehero.back.config.ResilienceConstants.INSTANCE_PROGRESSION_SERVICE;

@Slf4j
@Service
public class ProgressionService implements IProgressionService {

    private static Long ZERO = 0L;

    @Autowired
    private IStoryService storyService;
    @Autowired
    private IPersonService personService;
    @Autowired
    private IPageService pageService;
    @Autowired
    private ProgressionDao progressionDao;

    public Progression doProgressionAction(final Integer personId, final Integer storyId, final Integer pageId) {
        Assert.notNull(personId, "The personId parameter is mandatory !");
        Assert.notNull(storyId, "The storyId parameter is mandatory !");
        Assert.notNull(pageId, "The pageId parameter is mandatory !");

        if (personService.notExists(personId)) {
            throw new PersonNotFoundException(MessageFormatter.format("La personne avec l'id {} n'existe pas", personId).getMessage());
        }
        if (storyService.notExists(storyId)) {
            throw new StoryNotFoundException(MessageFormatter.format("L'histoire avec l'id {} n'existe pas", storyId).getMessage());
        }
        if (pageService.notExists(pageId)) {
            throw new PageNotFoundException(MessageFormatter.format("La page avec l'id {} n'existe pas", pageId).getMessage());
        }

        Optional<Progression> optionalProgression = progressionDao.findByPersonIdAndStoryId(personId, storyId);
        if (optionalProgression.isPresent()) {
            Progression progression = optionalProgression.get();

            // TODO: vérifier qu'on a le droit d'avancer sur cette page depuis la page précédente
            progression.setActualPageId(pageId);

            return progressionDao.save(progression);
        } else {
            throw new ProgressionNotFoundException("Aucune progression avec le personId " + personId + " et le storyId " + storyId);
        }
    }

    @CircuitBreaker(name = INSTANCE_PROGRESSION_SERVICE, fallbackMethod = "emptyProgressionList")
    public List<Progression> findByPersonId(final Integer personId) {
        Assert.notNull(personId, "The personId parameter is mandatory !");
        if (personService.notExists(personId)) {
            throw new PersonNotFoundException(MessageFormatter.format("La personne avec l'id {} n'existe pas", personId).getMessage());
        }

        return progressionDao.findByPersonId(personId);
    }

    @SuppressWarnings({"squid:UnusedPrivateMethod", "squid:S1172"})
    private List<Progression> emptyProgressionList(final Integer personId, final Throwable e) {
        return Collections.emptyList();
    }

    @CircuitBreaker(name = INSTANCE_PROGRESSION_SERVICE, fallbackMethod = "emptyProgression")
    public Optional<Progression> findByPersonIdAndStoryId(final Integer personId, final Integer storyId) {
        Assert.notNull(personId, "The personId parameter is mandatory !");
        Assert.notNull(storyId, "The storyId parameter is mandatory !");
        if (personService.notExists(personId)) {
            throw new PersonNotFoundException(MessageFormatter.format("La personne avec l'id {} n'existe pas", personId).getMessage());
        }
        if (storyService.notExists(storyId)) {
            throw new StoryNotFoundException(MessageFormatter.format("L'histoire avec l'id {} n'existe pas", storyId).getMessage());
        }

        return progressionDao.findByPersonIdAndStoryId(personId, storyId);
    }

    @SuppressWarnings({"squid:UnusedPrivateMethod", "squid:S1172"})
    private Optional<Page> emptyProgression(final Integer personId, final Integer storyId, final Throwable e) {
        return Optional.empty();
    }

    @CircuitBreaker(name = INSTANCE_PROGRESSION_SERVICE, fallbackMethod = "emptyProgressionCount")
    public Long countByStoryId(final Integer storyId) {
        Assert.notNull(storyId, "The storyId parameter is mandatory !");
        if (storyService.notExists(storyId)) {
            throw new StoryNotFoundException(MessageFormatter.format("L'histoire avec l'id {} n'existe pas", storyId).getMessage());
        }

        return progressionDao.countByStoryId(storyId);
    }

    @SuppressWarnings({"squid:UnusedPrivateMethod", "squid:S1172"})
    private Long emptyProgressionCount(final Integer storyId, final Throwable e) {
        return ZERO;
    }
}

File: back/src/main/java/com/tdevilleduc/urthehero/back/service/impl/ProgressionService.kt
package com.tdevilleduc.urthehero.back.service.impl

import com.tdevilleduc.urthehero.back.constant.ApplicationConstants
import com.tdevilleduc.urthehero.back.dao.ProgressionDao
import com.tdevilleduc.urthehero.back.exceptions.PageNotFoundException
import com.tdevilleduc.urthehero.back.exceptions.ProgressionNotFoundException
import com.tdevilleduc.urthehero.back.exceptions.StoryNotFoundException
import com.tdevilleduc.urthehero.back.exceptions.UserNotFoundException
import com.tdevilleduc.urthehero.back.model.Progression
import com.tdevilleduc.urthehero.back.service.IPageService
import com.tdevilleduc.urthehero.back.service.IProgressionService
import com.tdevilleduc.urthehero.back.service.IStoryService
import com.tdevilleduc.urthehero.back.service.IUserService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.slf4j.helpers.MessageFormatter
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.util.Assert
import java.util.*

@Service
class ProgressionService : IProgressionService {

    @Autowired
    private lateinit var storyService: IStoryService
    @Autowired
    private lateinit var userService: IUserService
    @Autowired
    private lateinit var pageService: IPageService
    @Autowired
    private lateinit var progressionDao: ProgressionDao

    override fun doProgressionAction(userId: Int, storyId: Int, newPageId: Int): Progression {
        Assert.notNull(userId, ApplicationConstants.CHECK_USERID_PARAMETER_MANDATORY)
        Assert.notNull(storyId, ApplicationConstants.CHECK_STORYID_PARAMETER_MANDATORY)
        Assert.notNull(newPageId, ApplicationConstants.CHECK_PAGEID_PARAMETER_MANDATORY)
        if (userService.notExists(userId)) {
            throw UserNotFoundException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_USER_DOESNOT_EXIST, userId).message)
        }
        if (storyService.notExists(storyId)) {
            throw StoryNotFoundException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_STORY_DOESNOT_EXIST, storyId).message)
        }
        if (pageService.notExists(newPageId)) {
            throw PageNotFoundException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_PAGE_DOESNOT_EXIST, newPageId).message)
        }
        val progression = this.findByUserIdAndStoryId(userId, storyId)
        // TODO: vérifier qu'on a le droit d'avancer sur cette page depuis la page précédente
        progression.actualPageId = newPageId
        return progressionDao.save(progression)
    }

    override fun findByUserId(userId: Int): MutableList<Progression> {
        Assert.notNull(userId, ApplicationConstants.CHECK_USERID_PARAMETER_MANDATORY)
        if (userService.notExists(userId)) {
            throw UserNotFoundException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_USER_DOESNOT_EXIST, userId).message)
        }
        return progressionDao.findByUserId(userId)
    }

    override fun findByUserIdAndStoryId(userId: Int, storyId: Int): Progression {
        val optional = progressionDao.findByUserIdAndStoryId(userId, storyId)
        return if (optional.isPresent) {
            optional.get()
        } else {
            throw ProgressionNotFoundException("Aucune progression avec le userId $userId et le storyId $storyId")
        }
    }

    override fun countUsersByStoryId(storyId: Int): Long {
        Assert.notNull(storyId, ApplicationConstants.CHECK_STORYID_PARAMETER_MANDATORY)
        if (storyService.notExists(storyId)) {
            throw StoryNotFoundException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_STORY_DOESNOT_EXIST, storyId).message)
        }
        return progressionDao.countByStoryId(storyId)
    }
}